samples/nrf9160/modem_shell/src/cloud/cloud_shell.c

```diff
@@ -0,0 +1,179 @@
+#endif
+
+BUILD_ASSERT(
+	IS_ENABLED(CONFIG_NRF_CLOUD_MQTT) &&
+	IS_ENABLED(CONFIG_NRF_CLOUD_CONNECTION_POLL_THREAD));
+
+#if defined(CONFIG_NRF_CLOUD_PGPS)
+static struct k_work notify_pgps_work;
+#endif
+
+static int cloud_shell_print_usage(const struct shell *shell, size_t argc, char **argv)
+{
+	int ret = 1;
+
+	if (argc > 1) {
+		mosh_error("%s: subcommand not found", argv[1]);
+		ret = -EINVAL;
+	}
+
+	shell_help(shell);
+
+	return ret;
+}
+
+#if defined(CONFIG_NRF_CLOUD_PGPS)
+static void notify_pgps(struct k_work *work)
+{
+	ARG_UNUSED(work);
+	int err;
+
+	err = nrf_cloud_pgps_notify_prediction();
+	if (err) {
+		mosh_error("Error requesting notification of prediction availability: %d", err);
+	}
+}
+#endif /* defined(CONFIG_NRF_CLOUD_PGPS) */
+
+static void nrf_cloud_event_handler(const struct nrf_cloud_evt *evt)
+{
+	int err = 0;
+
+	switch (evt->type) {
+	case NRF_CLOUD_EVT_TRANSPORT_CONNECTING:
+		mosh_print("NRF_CLOUD_EVT_TRANSPORT_CONNECTING");
+		break;
+	case NRF_CLOUD_EVT_TRANSPORT_CONNECTED:
+		mosh_print("NRF_CLOUD_EVT_TRANSPORT_CONNECTED");
+		break;
+	case NRF_CLOUD_EVT_READY:
+		mosh_print("NRF_CLOUD_EVT_READY");
+		break;
+	case NRF_CLOUD_EVT_TRANSPORT_DISCONNECTED:
+		mosh_print("NRF_CLOUD_EVT_TRANSPORT_DISCONNECTED");
+		break;
+	case NRF_CLOUD_EVT_ERROR:
+		mosh_print("NRF_CLOUD_EVT_ERROR");
+		break;
+	case NRF_CLOUD_EVT_SENSOR_DATA_ACK:
+		mosh_print("NRF_CLOUD_EVT_SENSOR_DATA_ACK");
+		break;
+	case NRF_CLOUD_EVT_FOTA_DONE:
+		mosh_print("NRF_CLOUD_EVT_FOTA_DONE");
+		break;
+	case NRF_CLOUD_EVT_RX_DATA:
+		mosh_print("NRF_CLOUD_EVT_RX_DATA");
+		if (((char *)evt->data.ptr)[0] == '{') {
+			/* Not A-GPS data */
+			break;
+		}
+#if defined(CONFIG_NRF_CLOUD_AGPS)
+		err = nrf_cloud_agps_process((char *)evt->data.ptr, evt->data.len);
+		if (!err) {
+			mosh_print("A-GPS data processed");
+#if defined(CONFIG_NRF_CLOUD_PGPS)
+			/* call us back when prediction is ready */
+			k_work_submit(&notify_pgps_work);
+#endif
+			/* data was valid; no need to pass to other handlers */
+			break;
+		}
+#endif
+#if defined(CONFIG_NRF_CLOUD_PGPS)
+		err = nrf_cloud_pgps_process(evt->data.ptr, evt->data.len);
+		if (err) {
+			mosh_error("Error processing P-GPS packet: %d", err);
+		}
+#else
+		if (err) {
+			mosh_print("Unable to process A-GPS data, error: %d", err);
+		}
+#endif
+		break;
+	case NRF_CLOUD_EVT_USER_ASSOCIATION_REQUEST:
+		mosh_print("NRF_CLOUD_EVT_USER_ASSOCIATION_REQUEST");
+		mosh_warn("Add the device to nRF Cloud and wait for it to reconnect");
+		break;
+	case NRF_CLOUD_EVT_USER_ASSOCIATED:
+		mosh_print("NRF_CLOUD_EVT_USER_ASSOCIATED");
+		break;
+	default:
+		mosh_error("Unknown nRF Cloud event type: %d", evt->type);
+		break;
+	}
+}
+
+static void cmd_cloud_connect(const struct shell *shell, size_t argc, char **argv)
+{
+	int err;
+
+	struct nrf_cloud_init_param config = {
+		.event_handler = nrf_cloud_event_handler,
+	};
+
+	err = nrf_cloud_init(&config);
+	if (err && (err != -EACCES)) {
```

Comment:
Doesn't this change the behavior with `-EACCES`? Earlier that resulted in the `mosh_print()` call, but now it would get handle by the `else` branch if P-GPS is enabled, which is not intended behavior. Either the `else` branch has to check that `err` is not `-EACCES`, or it may be easier to just add a static bool flag to store whether initialization has been done or not.

Reply:
flag added